pkg/cloud/aws/actuators/machine/utils_test.go

```diff
@@ -0,0 +1,123 @@
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"testing"
+)
+
+func init() {
+	// Add types to scheme
+	clusterv1.AddToScheme(scheme.Scheme)
+}
+
+func TestProviderConfigFromMachine(t *testing.T) {
+
+	providerConfig := &providerconfigv1.AWSMachineProviderConfig{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "awsproviderconfig.k8s.io/v1alpha1",
+			Kind:       "AWSMachineProviderConfig",
+		},
+		InstanceType: "testInstance",
+		AMI:          providerconfigv1.AWSResourceReference{ID: nil},
+		Tags: []providerconfigv1.TagSpecification{
+			{Name: "", Value: ""},
+		},
+		IAMInstanceProfile: &providerconfigv1.AWSResourceReference{ID: nil},
+		UserDataSecret:     &corev1.LocalObjectReference{Name: ""},
+		Subnet: providerconfigv1.AWSResourceReference{
+			Filters: []providerconfigv1.Filter{{
+				Name:   "tag:Name",
+				Values: []string{""},
+			}},
+		},
+		Placement: providerconfigv1.Placement{Region: "", AvailabilityZone: ""},
+		SecurityGroups: []providerconfigv1.AWSResourceReference{{
+			Filters: []providerconfigv1.Filter{{
+				Name:   "tag:Name",
+				Values: []string{""},
+			}},
+		}},
+	}
+
+	codec, err := providerconfigv1.NewCodec()
+	if err != nil {
+		t.Error(err)
+	}
+	encodedProviderConfig, err := codec.EncodeProviderConfig(providerConfig)
+	if err != nil {
+		t.Error(err)
+	}
+
+	machineClass := &clusterv1.MachineClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "openshift-cluster-api",
+			Name:      "testClass",
+		},
+		ProviderConfig: *encodedProviderConfig.Value,
+	}
+
+	testCases := []struct {
+		machine *clusterv1.Machine
+	}{
+		{
+			machine: &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "configFromSpecProviderConfigValue",
+					Namespace: "",
+					Labels: map[string]string{
+						"foo": "a",
+					},
+				},
+				TypeMeta: metav1.TypeMeta{
+					Kind: "Machine",
+				},
+				Spec: clusterv1.MachineSpec{
+					ProviderConfig: *encodedProviderConfig,
+				},
+			},
+		},
+		{
+			machine: &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "configFromClass",
+					Namespace: "",
+					Labels: map[string]string{
+						"foo": "a",
+					},
+				},
+				TypeMeta: metav1.TypeMeta{
+					Kind: "Machine",
+				},
+				Spec: clusterv1.MachineSpec{
+					ProviderConfig: clusterv1.ProviderConfig{
+						ValueFrom: &clusterv1.ProviderConfigSource{
+							MachineClass: &clusterv1.MachineClassRef{
+								ObjectReference: &corev1.ObjectReference{
+									Kind:      "MachineClass",
+									Name:      "testClass",
+									Namespace: "openshift-cluster-api",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	client := fake.NewFakeClient(machineClass)
+	for _, tc := range testCases {
+		decodedProviderConfig, err := ProviderConfigFromMachine(client, tc.machine)
+		if err != nil {
+			t.Error(err)
+		}
+		if decodedProviderConfig.InstanceType != "testInstance" {
```

Comment:
`reflect.DeepEqual` will do better so you don't have to test for concrete fields.